### DISCUSSION TITLE
Is it possible create background task wich should do something.

### DISCUSSION THREAD

**Discussion by heckad** [tokio-rs/tokio#2644] [General] [answered]
I want to print an app state every minute. To solve this problem, I need to do something that will work in the background and when I exit from a scope this background task should stop. 
For example

```
async fn some_func() {
    let monitoring  = BackgroundWorker::new(move || info!("Current state ..."));

    // Do something
    // When exit we in drop should stop background task.
}
```

How to do this or maybe we have a library yet?

**Comment by Darksonn** [comment, marked as answer]
The easiest way to kill the task when you exit the scope is to use [`remote_handle`][1] from the `futures-util` crate. Note that this requires your state to be share-able in some manner. If your state is currently not share-able, the [mini-redis][2] example has an illustration of a common way of sharing mutable state in [`db.rs`][3].
```rust
use tokio::time::{self, Duration};

async fn some_func(state: AppState) {
    let (job, _handle) = futures_util::future::FutureExt::remote_handle(monitor(state.clone()));
    tokio::spawn(job);

    // do stuff

    // _handle goes out of scope here, which kills the task
}

async fn monitor(state: AppState) {
    let mut interval = time::interval(Duration::from_secs(60)); // or delay_for
    loop {
        interval.tick().await;
        info!("App state is {:?}", state);
    }
}
```
As an alternative to `remote_handle`, there is also [`abortable`][4] from the `futures` crate, but you have to explicitly call `abort` when using this tool. If using `abortable`, consider defining a destructor on some value and calling `abort()` in the destructor, as you otherwise risk leaking the task in case of panics.

---

As an alternative, you can use the [`select!`][5] macro instead of spawning.
```rust
use tokio::time::{interval, Duration};

async fn some_func(state1: AppState) {
    let state2 = state1.clone();
    tokio::select! {
        _ = do_stuff(state1) => {},
        _ = monitor(state2) => {},
    }
}

async fn do_stuff(state: AppState) {
    // ...
}

async fn monitor(state: AppState) {
    let mut interval = interval(Duration::from_secs(60));
    loop {
        interval.tick().await;
        info!("App state is {:?}", state);
    }
}
```

[1]: https://docs.rs/futures-util/0.3/futures_util/future/trait.FutureExt.html#method.remote_handle
[2]: https://github.com/tokio-rs/mini-redis/
[3]: https://github.com/tokio-rs/mini-redis/blob/master/src/db.rs
[4]: https://docs.rs/futures/0.3/futures/future/fn.abortable.html
[5]: https://docs.rs/tokio/0.2/tokio/macro.select.html

**Reply by heckad** [reply]
How to use `abortable` task with `tokio::spawn`? It would be nice to add an example to the documentation.

**Reply by Darksonn** [reply]
Similar to `remote_handle`.
```rust
use futures::future::abortable;

let (job, handle) = abortable(task());
tokio::spawn(job);

...

handle.abort();
```

**Reply by heckad** [reply]
Thank you very much!
